### PR TITLE
add calendar reminder for l2 bridging

### LINKED ISSUE #8663: Calendar reminders for L2 bridging to mainnet
## Abstract

There is many bridging events that require about a week of waiting between bridging from L2 and claiming in mainnet. Create a calendar reminder entry for them.

## Motivation
https://x.com/superphiz/status/1841188404052312142

Well .... not to forget or have to use google calendar or something. Let the reminder be auto-created when the event is seen onchain.

## Specification

Do it

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #8663 

Added L2->L1 bridge claim reminders for Base, Optimism, and Arbitrum. 

Withdrawal periods:
- Base: [>= 7 days](https://docs.optimism.io/builders/app-developers/bridging/messaging#for-l2-to-l1-transactions-1)
- Optimism: [>= 7 days](https://docs.optimism.io/builders/app-developers/bridging/messaging#for-l2-to-l1-transactions-1) 
- Arbitrum:  [>= 7 days](https://docs.arbitrum.io/arbitrum-bridge/quickstart#withdraw-eth-or-erc-20-tokens-from-child-chain-to-parent-chain)
- [Not added] Scroll: [4 hours](https://docs.scroll.io/en/user-guide/bridge/)

### CODE REVIEW

File: rotkehlchen/tasks/calendar.py [hunk 1]
```diff
@@ -410,6 +422,60 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                    name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
+                    event=bridge_event,
+                    color=BRIDGE_CALENDAR_COLOR,
+                    counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
+                    timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),
+                    description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} is ready to claim on Ethereum',  # noqa: E501
+                )
+
+                if entry_id is not None:
+                    bridge_calendar_entries.append(entry_id)
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:468]: use the walrus operator here

```suggestion
                if (entry_id := self.create_or_update_calendar_entry_from_event(
                    name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
                    event=bridge_event,
                    color=BRIDGE_CALENDAR_COLOR,
                    counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
                    timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),
                    description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} is ready to claim on Ethereum',  # noqa: E501
                )) is not None:
                    bridge_calendar_entries.append(entry_id)
```

File: rotkehlchen/tasks/calendar.py [hunk 2]
```diff
@@ -410,6 +422,60 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
             error_msg='Failed to add the airdrop claim reminders',
         )
 
+    def maybe_create_l2_bridging_reminder(self) -> None:
+        """Creates calendar reminders for L2 bridge deposit claims.
+        Tracks deposits on Base, Optimism, and Arbitrum networks.
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:427]: can you state in a comment that base, op and arb have a 1 week withdrawal time?

File: rotkehlchen/tasks/calendar.py [hunk 3]
```diff
@@ -410,6 +422,60 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                        event_types=[HistoryEventType.DEPOSIT],
+                        event_subtypes=[HistoryEventSubType.BRIDGE],
+                    ),
+                ))
+
+        now = ts_now_in_ms()
+        bridge_calendar_entries: list[int] = []
+        for bridge_event in bridge_events:
+            if now - bridge_event.timestamp > seven_days_in_ms:
+                continue
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:455]: I wouldn't use `ts_sec_to_ms` here since is not a timestamp but a duration and it can be confusing. I would just do `WEEK_IN_SECONDS * 1000` here

File: rotkehlchen/tasks/calendar.py [hunk 4]
```diff
@@ -22,20 +25,29 @@
...
-from rotkehlchen.utils.misc import ts_now
+from rotkehlchen.utils.misc import ts_ms_to_sec, ts_now, ts_now_in_ms, ts_sec_to_ms
 from rotkehlchen.utils.mixins.customizable_date import CustomizableDateMixin
 
 logger = logging.getLogger(__name__)
 log = RotkehlchenLogsAdapter(logger)
 ENS_CALENDAR_COLOR: Final = deserialize_hex_color_code('5298FF')
 CRV_CALENDAR_COLOR: Final = deserialize_hex_color_code('5bf054')
 AIRDROP_CALENDAR_COLOR: Final = deserialize_hex_color_code('ffd966')
+BRIDGE_CALENDAR_COLOR: Final = deserialize_hex_color_code('eeeeee')
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:50]: why this color. Can you ask the frontend if they have a pallete for colors to use in calendars?

File: rotkehlchen/tests/unit/decoders/test_superchain_bridge.py [hunk 5]
```diff
@@ -592,13 +583,11 @@ def test_withdraw_eth_base_to_ethereum_bridge(base_inquirer, base_accounts):
 @pytest.mark.parametrize('base_accounts', [['0x779e4b47c3Dea5689233821dFcf429E0485eF116']])
 def test_withdraw_erc20_base_to_ethereum_bridge(base_inquirer, base_accounts):
     evmhash = deserialize_evm_tx_hash('0xbe4e54e77cb700f2755b236d7823295129c4d5e22fdc87df8058274bc0fefab1')  # noqa: E501
-    events, _ = get_decoded_events_of_transaction(evm_inquirer=base_inquirer, tx_hash=evmhash)
-    gas_amount, deposit_amount = '0.000029210771559657', '49352.072702'
-    user_address = base_accounts[0]
+    (events, _), gas_amount, deposit_amount, user_address, timestamp = get_decoded_events_of_transaction(evm_inquirer=base_inquirer, tx_hash=evmhash), '0.000029210771559657', '49352.072702', base_accounts[0], TimestampMS(1729293119000)  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_superchain_bridge.py:586]: why did you touch them? It's good to move the timestamp but I would leave `events, _ = get_decoded_events_of_transaction(evm_inquirer=base_inquirer, tx_hash=evmhash)` in its own line. If you don't need to modify this file revert the changes since it adds noise to the commit

File: rotkehlchen/tasks/calendar.py [hunk 6]
```diff
@@ -410,6 +422,60 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                    bridge_calendar_entries.append(entry_id)
+            except UnknownAsset:
+                log.exception(f'Unable to add reminder for brige event with identifier, {bridge_event.identifier}')  # noqa: E501
+                continue
+
+        self.maybe_create_reminders(
+            secs_befores=[0],
+            calendar_identifiers=bridge_calendar_entries,
+            error_msg='Failed to add the bridge claim reminders',
+        )
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:477]: Do you think we need a delete of the reminder if the user claims the bridge? I'm thinking that the only scenario is the transactions getting decoded before the task for the reminders runs since you can't withdraw before... Maybe is not needed

File: rotkehlchen/tests/unit/test_calendar_reminders.py [hunk 7]
```diff
@@ -260,3 +262,126 @@ def mock_requests_get(url: str, timeout: int = 0, headers: dict | None = None):
...
+@pytest.mark.freeze_time('2023-06-30 15:45:00 GMT')
+@pytest.mark.parametrize('arbitrum_one_accounts', [[
+    '0x5EA45c8E36704d7F4053Bb0e23cDd96E4d8b80F7',
+]])
+def test_l2_arbitrum_bridge_claim_reminder(arbitrum_one_accounts, arbitrum_one_inquirer, database):
+    evmhash = deserialize_evm_tx_hash('0xdb8e29f27a7b7b416f168e8135347703268a142b6776503e26419dbfc43bcabf')  # noqa: E501
+    events, _ = get_decoded_events_of_transaction(
+        tx_hash=evmhash,
+        evm_inquirer=arbitrum_one_inquirer,
+    )
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_calendar_reminders.py:277]:
```suggestion
    events, _ = get_decoded_events_of_transaction(
      tx_hash=deserialize_evm_tx_hash('0xdb8e29f27a7b7b416f168e8135347703268a142b6776503e26419dbfc43bcabf')  # noqa: E501,
        evm_inquirer=arbitrum_one_inquirer,
    )
```

File: rotkehlchen/tests/unit/test_calendar_reminders.py [hunk 8]
```diff
@@ -260,3 +262,126 @@ def mock_requests_get(url: str, timeout: int = 0, headers: dict | None = None):
...
+    assert reminders[0].event_id == calendar_entry.identifier
+    assert reminders[0].secs_before == 0
+
+
+@pytest.mark.vcr(filter_query_parameters=['apikey'])
+@pytest.mark.freeze_time('2024-10-18 07:24:00 GMT')
+@pytest.mark.parametrize('base_accounts', [[
+    '0x6730b1Df17E50217777EeE475E34815964e3BFb2',
+]])
+def test_l2_base_bridge_claim_reminder(base_accounts, base_inquirer, database):
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/test_calendar_reminders.py:355]: it seems that you can abstract those 3 tests into 1 with a for loop, don't you think? You would have 3 inquirers loaded and query 3 chains but that is fine

File: rotkehlchen/tasks/calendar.py [hunk 9]
```diff
@@ -410,6 +422,60 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
         )
 
+    def maybe_create_l2_bridging_reminder(self) -> None:
+        """Creates calendar reminders for L2 bridge deposit claims.
+        Tracks deposits on Base, Optimism, and Arbitrum networks.
+        """
+        bridge_events: list[EvmEvent] = []
+        seven_days_in_ms = ts_sec_to_ms(Timestamp(WEEK_IN_SECONDS))
+        locations_to_counterparties = [
+            (Location.BASE, CPT_BASE),
```
Comment (yabirgb) [inline rotkehlchen/tasks/calendar.py:432]: just a nitpick but make this a tuple instead of a list

File: rotkehlchen/tasks/calendar.py (hunk 6 above)
Comment (prettyirrelevant) [reply rotkehlchen/tasks/calendar.py:477]: they're created with `auto_delete=True` so they get deleted here 

https://github.com/prettyirrelevant/rotki/blob/477617d01c94e3b7d72d684d0d2b45e322f9c565/rotkehlchen/tasks/manager.py#L880

File: rotkehlchen/tests/unit/decoders/test_superchain_bridge.py (hunk 5 above)
Comment (prettyirrelevant) [reply rotkehlchen/tests/unit/decoders/test_superchain_bridge.py:586]: they're nitpicks from a previous pr. lefteris said to add them in a follow up pr.

File: rotkehlchen/tests/unit/test_calendar_reminders.py [hunk 10]
```diff
@@ -260,3 +263,60 @@ def mock_requests_get(url: str, timeout: int = 0, headers: dict | None = None):
...
+                tx_hash=deserialize_evm_tx_hash(tx_hash),
+                evm_inquirer=inquirer,
+            )
+
+            calendar_db = DBCalendar(database)
+            all_calendar_entries = calendar_db.query_calendar_entry(CalendarFilterQuery.make())
+            assert all_calendar_entries['entries_total'] == expected_entries
+
+            reminder_creator = CalendarReminderCreator(database=database, current_ts=ts_now())
+            reminder_creator.maybe_create_l2_bridging_reminder()
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/tests/unit/test_calendar_reminders.py]: yes

File: rotkehlchen/tasks/calendar.py [hunk 11]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                    filter_query=EvmEventFilterQuery.make(
+                        and_op=True,
+                        location=location,
+                        counterparties=[counterparty],
+                        event_types=[HistoryEventType.DEPOSIT],
+                        event_subtypes=[HistoryEventSubType.BRIDGE],
+                    ),
+                ))
+
+        now = ts_now_in_ms()
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:450]: you don't reuse this. Just don't use a variable

File: rotkehlchen/tasks/calendar.py [hunk 12]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
+                        event=bridge_event,
+                        color=BRIDGE_CALENDAR_COLOR,
+                        counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
+                        timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),  # noqa: E501
+                        description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} is ready to claim on Ethereum',  # noqa: E501
+                )) is not None:
+                    bridge_calendar_entries.append(entry_id)
+            except UnknownAsset:
+                log.exception(f'Unable to add reminder for brige event with identifier, {bridge_event.identifier}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:467]: They are all evm events no? Isn't it better to just use evm_event.tx_hash + chain id?

```suggestion
                log.exception(f'Unable to add reminder for brige event with identifier, {bridge_event.identifier}')  # noqa: E501
```

File: rotkehlchen/tasks/calendar.py [hunk 13]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                continue
+
+            try:
+                if (entry_id := self.create_or_update_calendar_entry_from_event(
+                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
+                        event=bridge_event,
+                        color=BRIDGE_CALENDAR_COLOR,
+                        counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
+                        timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),  # noqa: E501
+                        description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} is ready to claim on Ethereum',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:463]: Shouldn't they always have a symbol? These should be evm tokens.

```suggestion
                        description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol()} is ready to claim on Ethereum',  # noqa: E501
```

File: rotkehlchen/tasks/calendar.py [hunk 14]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+
+        now = ts_now_in_ms()
+        bridge_calendar_entries: list[int] = []
+        for bridge_event in bridge_events:
+            if now - bridge_event.timestamp > WEEK_IN_SECONDS * 1000:
+                continue
+
+            try:
+                if (entry_id := self.create_or_update_calendar_entry_from_event(
+                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:458]: as in the other comment

```suggestion
                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol()} bridge deposit on Ethereum',  # noqa: E501
```

File: rotkehlchen/tasks/calendar.py [hunk 15]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                        counterparties=[counterparty],
+                        event_types=[HistoryEventType.DEPOSIT],
+                        event_subtypes=[HistoryEventSubType.BRIDGE],
+                    ),
+                ))
+
+        now = ts_now_in_ms()
+        bridge_calendar_entries: list[int] = []
+        for bridge_event in bridge_events:
+            if now - bridge_event.timestamp > WEEK_IN_SECONDS * 1000:
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:453]: Why not jsut use ts_now()?

```suggestion
            if ts_now() - bridge_event.timestamp > WEEK_IN_SECONDS:
```

File: docs/changelog.rst [hunk 16]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :feature:`8663` Rotki will now create calendar reminders for claiming bridge withdrawals from Layer 2 networks (Base, Optimism, Arbitrum) after the 7-day waiting period.
```
Comment (LefterisJP) [inline docs/changelog.rst:5]:
```suggestion
* :feature:`8663` rotki will now create calendar reminders for claiming bridge withdrawals from Layer 2 networks (Base, Optimism, Arbitrum) after the 7-day waiting period.
```

File: rotkehlchen/tests/unit/test_calendar_reminders.py [hunk 17]
```diff
@@ -260,3 +263,62 @@ def mock_requests_get(url: str, timeout: int = 0, headers: dict | None = None):
...
+            new_calendar_entries = calendar_db.query_calendar_entry(CalendarFilterQuery.make())
+            assert new_calendar_entries['entries_found'] == expected_entries
+
+            calendar_entry = new_calendar_entries['entries'][-1]
+            bridge_event = events[-1]
+            assert calendar_entry == CalendarEntry(
+                identifier=idx + 1,
+                name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} bridge deposit on Ethereum',  # noqa: E501
+                timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),
+                description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol_or_name()} is ready to claim on Ethereum',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_calendar_reminders.py:312]: Generally here and everywhere else in the PR. Use only symbol

```suggestion
                description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.symbol()} is ready to claim on Ethereum',  # noqa: E501
```

File: rotkehlchen/tasks/calendar.py (hunk 15 above)
Comment (prettyirrelevant) [reply rotkehlchen/tasks/calendar.py:453]: `bridge_event.timestamp` is in milliseconds. also why `now = ts_now_in_ms()` is defined outside of the for-loop.

File: rotkehlchen/tasks/calendar.py (hunk 11 above)
Comment (prettyirrelevant) [reply rotkehlchen/tasks/calendar.py:450]: image

File: rotkehlchen/tasks/calendar.py [hunk 18]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.resolve_to_asset_with_symbol().symbol} bridge deposit on Ethereum',  # noqa: E501
+                        event=bridge_event,
+                        color=BRIDGE_CALENDAR_COLOR,
+                        counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
+                        timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),  # noqa: E501
+                        description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.resolve_to_asset_with_symbol().symbol} is ready to claim on Ethereum',  # noqa: E501
+                )) is not None:
+                    bridge_calendar_entries.append(entry_id)
+            except UnknownAsset:
+                log.exception(f'Unable to add reminder for brige event with hash {bridge_event.identifier} on {bridge_event.location.name}')  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:467]: As I said in my last review. You need the tx hash. The DB identifier means nothing.

```suggestion
                log.exception(f'Unable to add reminder for bridge event with hash {bridge_event.tx_hash} on {bridge_event.location.name}')  # noqa: E501
```

File: rotkehlchen/tasks/calendar.py [hunk 19]
```diff
@@ -410,6 +422,57 @@ def maybe_create_airdrop_claim_reminder(self) -> None:
...
+                continue
+
+            try:
+                if (entry_id := self.create_or_update_calendar_entry_from_event(
+                        name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.resolve_to_asset_with_symbol().symbol} bridge deposit on Ethereum',  # noqa: E501
+                        event=bridge_event,
+                        color=BRIDGE_CALENDAR_COLOR,
+                        counterparty=bridge_event.counterparty,  # type: ignore[arg-type]  # counterparty is always present
+                        timestamp=ts_ms_to_sec(TimestampMS(bridge_event.timestamp + WEEK_IN_SECONDS)),  # noqa: E501
+                        description=f'Bridge deposit of {bridge_event.balance.amount} {bridge_event.asset.resolve_to_asset_with_symbol().symbol} is ready to claim on Ethereum',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tasks/calendar.py:463]: please do the resolution only in a single place in the function. It's a heavy operation.

File: rotkehlchen/tests/unit/test_calendar_reminders.py [hunk 20]
```diff
@@ -260,3 +263,62 @@ def mock_requests_get(url: str, timeout: int = 0, headers: dict | None = None):
...
+            expected_entries += 1
+
+            new_calendar_entries = calendar_db.query_calendar_entry(CalendarFilterQuery.make())
+            assert new_calendar_entries['entries_found'] == expected_entries
+
+            calendar_entry = new_calendar_entries['entries'][-1]
+            bridge_event = events[-1]
+            assert calendar_entry == CalendarEntry(
+                identifier=idx + 1,
+                name=f'Claim {bridge_event.balance.amount} {bridge_event.asset.resolve_to_asset_with_symbol().symbol} bridge deposit on Ethereum',  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_calendar_reminders.py:310]: same as I said in the actual production code, though it matters less here since it's a test

**Review by LefterisJP — APPROVED** [review]
lgtm
